### ISSUE TITLE
allow text copy from main window

### ISSUE THREAD

**Issue by caver456** [ncssar/radiolog#568]
allow copy from main table and from team tables

- [x] make the cell editable, which allows drag-select, but disallow changes to content
- [x] don't clear all the text on edit
- [x] Ctrl-C during edit should copy selected text to system clipboard
- [x] exit edit and clear focus when any key besides Ctrl-C is pressed during edit
- [x] when any other key is pressed during edit, pass that key to main window, i.e. '1' during edit should open a NED for Team 1
- [x] get rid of dotted box (focus indicator) on top left cell after exiting edit
- [ ] ~~see if the first mouse-down can actually start the drag-select, instead of needing to double-left-click first to select the cell and start the editor~~
- [x] on first mouse press over a cell (left or right), open its editor with all text selected; then, allow drag-select in that cell
- [x] add right-click context menu with two items (copy, as alternative to Ctrl-C; and Amend this entry (#324))
- [x] if either context menu is open (table or editor), Ctrl-C should close context the menu (as well as copying to clipboard)
- [x] disable or override OS context menu that shows up on right-click when editor was already open - it shows actions for cut, paste, and delete, which should be blocked since the text can't be modified (and is reverted when focus is lost anyway)
- [x] apply all these features to team tables as well as the main table
- [x] unselect all and clear focus on next action outside the same table, e.g. clicking in different table
- [x] profile load of a large data set with this code vs previous version
- [x] see what happens when fleetsync data comes in at various stages of copy operations

**Comment by RadiosPRN** [comment]
+1 on this one. Would especially be nice to be able to copy the coordinate field. We do not use the sartopo locators from radiolog because typically we have live tracking with team members using the app but every once in a while it can be helpful to copy over the coordinates from radiolog to see a location we don't have a live track for and copy/paste would speed that up.

**Comment by caver456** [comment]
After some googling, this looks surprisingly tricky.  [This link](https://stackoverflow.com/questions/28223691) seems to have some good detail.

**Comment by caver456** [comment]
The delegate route (as above) may not be that terrible, but it seems pretty heavy-weight to add delegates for every cell.  Or can it be done on-the-fly, after a left or right click in the table?

One option that could definitely be done on-the-fly is to create a context menu on demand, just like for team tabs.  But, this would only allow selection of the entire cell.  Could trigger this context menu on left-click as well as right-click, in case the user is expecting to be able to do a left-click-drag-select over some text in the cell.

How important is it to allow partial cell (left-click-and-drag)?

**Comment by caver456** [comment]
Good progress, based on these posts:

Part one: making the cell editable without clearing all the text, by making sure a real value is returned when the edit role is requested:
https://www.qtcentre.org/threads/28555-QTableView-line-edit-clears-the-text-on-edit
https://www.qtcentre.org/threads/30013-TableView-Model-Display-Question

Part two: disallow changes while the cell is editable, using a delegate to capture keypress events:
https://stackoverflow.com/questions/65232193
https://stackoverflow.com/questions/39290017

**Comment by caver456** [comment]
Ctrl-B works as expected:
- right-click on cell when nothing was previously selected: context menu opens; shortcut copies to clipboard and closes context menu
- left-click on cell when nothing was previously selected: editor opens with all text selected; shortcut copies to clipboard
- left-click on cell when nothing was previously selected: editor opens and context menu opens; shortcut copied to clipboard and closes context menu

Basically, the shortcut should always copy to clipboard, and if a context menu is open, it should be closed.

This works for Ctrl-B.  But, when using Ctrl-C, it only copies to clipboard if a context menu is open.  If no context menu is open, the action does not appear to fire.

**Comment by caver456** [comment]
QEvent.ShortcutOverride (event type) was the ticket - read the docs [here](https://doc.qt.io/qt-6/qevent.html) which really clarified everything.

This was already being done in the code to prevent the system-level Ctrl-Z behavior.

```
		t=event.type()
		if(t==QEvent.ShortcutOverride and
				event.modifiers()==Qt.ControlModifier and
				event.key()==Qt.Key_C):
			# rprint('Ctrl-C detected')
			self.parent.copyText()
```

This does result in two calls to copyText when both the QAction and the ShortcutOverride event take effect on the same Ctrl-C keypress, but, that's not really a problem.

**Comment by caver456** [comment]
The 'final' item - applying the changes to the main tableView as well - is actually pretty hefty.  It requires promoting tableView from QTableView to CustomTableView (if it can be the same as the team tables), and since it's defined in Qt Designer, pyuic5 automatically adds a line 'from customTableView inlcude CustomTableView'.  Since the classes are currently defined in radiolog.py, and would need to be moved to a separate file to accomodate the above, there are follow-on requirements so that e.g. rprint can be called from the customTableView.py.  That seems a bit complex.  And, the locators issue is becoming a high priority, so, remaining work on this issue may be postponed for a while.

**Comment by caver456** [comment]
Possible solution to the Qt Designer promotion import problem: in Qt Designer, change the header file name for the custom widget class to 'radiolog.h' - that way, the line that gets added is 'from radiolog import CustomTableView'.  This might look like a circular import, but, it appears to be fine because only the one class is being imported a second time - not the whole file.  Run through all the checklist items again for this issue from the top, including profiling.

**Comment by caver456** [comment]
Reconsidering the passing of keystroke during edit to the top window.  It's a dilemma - either passing the first keystroke to the top, or, killing the first keystroke and exiting edit mode, both can cause confusion if the operator thinks they are modifying the text in the cell, and starts typing quickly.

Example: a cell says 'You are slow'.  Operator drag-selects 'slow' and quickly types 'fast' thinking they are editing the text.
Option 1 - first keystroke is passed to the top: the first key 'f' is a hotkey which opens a NED From, then 'irst' are used in the callsign, resulting in 'Team irst'
Option 2 - first keystroke is not passed to the top: the second key 'a' is also a hotkey so opens a NED TO All Teams with focus in the message field, which gets 'st'.

Which is better?  Definitely gray area.

Let's stick with the previous plan to pass the first keystroke to the top.  If you assume that the first keypress was intentional, i.e. '101' to add a new entry from team 101, then this makes sense.  If the first keypress is unintentional, then it doesn't make sense.  Time will tell and either way is hopefully not critical.  This could change later if needed.

**Comment by caver456** [comment]
@RadiosPRN - would you be available to download a beta install of this feature, test it out and let me know if it's user-friendly enough and intuitive, and if you have any suggestions?  It would be good to get another set of eyes on it before actually releasing it.  Let me know - I haven't built a beta release yet (actually haven't tried doing so before so that may take a few days).

**Comment by RadiosPRN** [comment]
Yes I would be happy to do that. Not sure on time frame because of some other things I have going on.

**Comment by caver456** [comment]
Great.  Will let you know when it's set.

**Comment by caver456** [comment]
Some load time profiling results: 
228 lines, 23kB

3.3.0:
~5.2 sec until progress box closes
~15 sec total until load dialog closes (~10 sec lag after progress box closes)

current dev code:
~6.5 sec until progress box closes (~1.3 sec slower, 25% slower)
~17 sec total until load dialog closes (~10.5 sec lag after progress box closes)

Copying and context menus after load in the current dev code don't seem to have any lag at all.

Need to test with a bigger data set...

**Comment by caver456** [comment]
@RadiosPRN ok that was easier than I thought, fortunately.  Just make a new release from the GUI, and the actions starts.  Unfortunately it bypasses bumpver, so the version number in the beta release (installer executable name, window title, transcript text, etc) all still say 3.3.0, but that's not a huge deal.

Please download and run the installer (radiolog-3.3.0-setup.zip) at this link: https://github.com/ncssar/radiolog/releases/tag/3.3.2beta

It will clobber any existing install, if that works for you.  It doesn't touch your working dir or config dir.  After testing, you can reinstall the latest stable version (3.3.1).  If you want to have a separate install dir anyway, the installer lets you specify what dir you want to install to.

Anyway, make some entries, try copying stuff, try to break it, and let me know what you think. Thanks

**Comment by RadiosPRN** [comment]
Initial testing on my personal computer seems to be working well. I have not used it yet while utilizing fleetsync data. I did notice a bug while playing with it which is probably a separate issue but when clicking more than one quick text button it adds a space before the last letter of every phrase except the first one entered which then does not change the Team Status button to the correct option.

**Comment by caver456** [comment]
thanks for the catch.  Reopened #506 - guess I buggered it up somehow!

**Comment by caver456** [comment]
profiling shows about 25-30% slower total lag time (25-30% slower during progress box, and 25-30% slower after progress box closes) vs 3.3.1 on an 887-entry radio log, created by copying a 223-line radio log csv into itself three times.  No noticeable lag on new fleetsync calls or new calls by typing callsign.  887 is probably a reasonable full-day operation entry count.  So, rather than digging into the lag reduction now, let's just say the increased lag is not a big deal for now and should not delay merging this branch.

**Comment by caver456** [comment]
Fixed this bug in the master branch, as well as another cleanup item in
this 'copy' feature.  Those will probably both be released in 3.4.0 in the
next few days.  Thanks again for the catch!
